File: Minimal/src/Brontide.Minimal.Binding/Portable/PortableShapeCatalog.fs
namespace Brontide.Minimal.Binding.Portable

/// The Shape floor's built-in references, which no contract needs to redeclare.
[<RequireQualifiedAccess>]
module PortableBuiltInShapes =

    let private literal name version =
        match PortableShapeRef.tryCreate name version with
        | Ok reference -> reference
        // A built-in outside the portable profile would be a defect in this module, not a
        // condition any peer can present, so it is not part of the portable failure model.
        | Error _ -> invalidOp $"'{name}@{version}' is not a portable Shape reference."

    let text = literal "Text" 1
    let boolean = literal "Boolean" 1
    let signed64 = literal "Integer.Signed64" 1
    let decimal = literal "Decimal" 1
    let bytes = literal "Bytes" 1
    let unit = literal "Unit" 1

    let declarations =
        [ { Reference = text; Body = ScalarBody PortableScalar.Text }
          { Reference = boolean; Body = ScalarBody PortableScalar.Boolean }
          { Reference = signed64; Body = ScalarBody PortableScalar.Signed64 }
          { Reference = decimal; Body = ScalarBody PortableScalar.Decimal }
          { Reference = bytes; Body = ScalarBody PortableScalar.Bytes }
          { Reference = unit; Body = UnitBody } ]

/// The Shape graph of one established contract, plus the projection rule that decides when a
/// version difference is additive.
///
/// An unresolvable reference fails as unsupported-contract rather than as a decode refusal, because
/// an unknown Shape, Fragment, or dependency is a contract fact and not a byte-level one.
[<StructuralEquality; NoComparison>]
type ShapeCatalog =
    private
        { Declarations: Map<PortableShapeRef, ShapeDeclaration>
          FragmentDeclarations: Map<PortableFragmentRef, FragmentDeclaration>
          Bounds: PortableLimits }

[<RequireQualifiedAccess>]
module ShapeCatalog =

    let limits catalog = catalog.Bounds

    let tryShape reference catalog = Map.tryFind reference catalog.Declarations

    let shape reference catalog : PortableResult<ShapeDeclaration> =
        match tryShape reference catalog with
        | Some declaration -> Ok declaration
        | None ->
            unsupportedContract
                "unknown-shape"
                $"Shape {PortableShapeRef.text reference} is not part of the established contract."

    let tryFragment reference catalog = Map.tryFind reference catalog.FragmentDeclarations

    let fragment reference catalog : PortableResult<FragmentDeclaration> =
        match tryFragment reference catalog with
        | Some declaration -> Ok declaration
        | None ->
            unsupportedContract
                "unknown-fragment"
                $"Fragment {PortableFragmentRef.text reference} is not part of the established contract."

    let private fieldsOf body =
        match body with
        | RecordBody(_, fields) -> fields
        | SequenceBody _
        | ChoiceBody _
        | ScalarBody _
        | UnitBody -> []

    let private requireUniqueFieldNames (owner: string) (fields: FieldDeclaration list) =
        let rec walk seen remaining =
            match remaining with
            | [] -> Ok()
            | (field: FieldDeclaration) :: tail ->
                match PortableMemberToken.tryCreate field.Name with
                | Error error -> Error error
                | Ok _ ->
                    if Set.contains field.Name seen then
                        malformed "duplicate-field" $"{owner} declares field '{field.Name}' more than once."
                    else
                        walk (Set.add field.Name seen) tail

        walk Set.empty fields

    let private addShape (declarations: Map<PortableShapeRef, ShapeDeclaration>) (declaration: ShapeDeclaration) =
        if Map.containsKey declaration.Reference declarations then
            malformed "duplicate-shape" $"Shape {PortableShapeRef.text declaration.Reference} is declared more than once."
        else
            requireUniqueFieldNames $"Shape {PortableShapeRef.text declaration.Reference}" (fieldsOf declaration.Body)
            |> Result.map (fun () -> Map.add declaration.Reference declaration declarations)

    let private addFragment (declarations: Map<PortableFragmentRef, FragmentDeclaration>) (declaration: FragmentDeclaration) =
        if Map.containsKey declaration.Reference declarations then
            malformed
                "duplicate-fragment"
                $"Fragment {PortableFragmentRef.text declaration.Reference} is declared more than once."
        else
            requireUniqueFieldNames $"Fragment {PortableFragmentRef.text declaration.Reference}" declaration.Fields
            |> Result.map (fun () -> Map.add declaration.Reference declaration declarations)

    let private requireResolvable (document: ContractDocument) catalog =
        let resolveShape reference = shape reference catalog |> Result.map ignore
        let resolveFragment reference = fragment reference catalog |> Result.map ignore

        portable {
            for declaration in document.Shapes do
                match declaration.Body with
                | RecordBody(_, fields) -> do! fields |> iterate (fun field -> resolveShape field.Shape)
                | SequenceBody item -> do! resolveShape item
                | ChoiceBody alternatives ->
                    do! alternatives |> iterate (fun alternative -> resolveShape alternative.Shape)
                | ScalarBody _
                | UnitBody -> ()

            for declaration in document.Fragments do
                do! resolveShape declaration.HostShape
                do! declaration.Fields |> iterate (fun field -> resolveShape field.Shape)

            for operation in document.Operations do
                do! resolveShape operation.InputShape
                do! resolveShape operation.ResultShape
                do! resolveShape operation.DetailShape
                do! operation.RequiredFragments |> iterate resolveFragment
        }

    let fromContract (document: ContractDocument) : PortableResult<ShapeCatalog> =
        let seedShapes =
            PortableBuiltInShapes.declarations
            |> List.map (fun declaration -> declaration.Reference, declaration)
            |> Map.ofList

        let rec foldShapes declarations remaining =
            match remaining with
            | [] -> Ok declarations
            | head :: tail -> addShape declarations head |> Result.bind (fun next -> foldShapes next tail)

        let rec foldFragments declarations remaining =
            match remaining with
            | [] -> Ok declarations
            | head :: tail -> addFragment declarations head |> Result.bind (fun next -> foldFragments next tail)

        portable {
            let! shapes = foldShapes seedShapes document.Shapes
            let! fragments = foldFragments Map.empty document.Fragments

            let catalog =
                { Declarations = shapes
                  FragmentDeclarations = fragments
                  Bounds = document.Limits }

            do! requireResolvable document catalog
            return catalog
        }

    /// Decides whether the later Shape version differs from the earlier one only additively: new
    /// optional fields. A new required field, a removed field, a changed field Shape, or a changed
    /// cardinality is not additive.
    let isAdditiveOver later earlier catalog =
        if PortableShapeRef.name later <> PortableShapeRef.name earlier
           || PortableShapeRef.version later < PortableShapeRef.version earlier then
            false
        elif later = earlier then
            true
        else
            match tryShape later catalog, tryShape earlier catalog with
            | Some laterShape, Some earlierShape ->
                match laterShape.Body, earlierShape.Body with
                | RecordBody(laterPolicy, laterFields), RecordBody(earlierPolicy, earlierFields) ->
                    let laterByName =
                        laterFields |> List.map (fun field -> field.Name, field) |> Map.ofList

                    let earlierNames = earlierFields |> List.map (fun field -> field.Name) |> Set.ofList

                    laterPolicy = earlierPolicy
                    && earlierFields
                       |> List.forall (fun field ->
                           match Map.tryFind field.Name laterByName with
                           | Some candidate -> candidate = field
                           | None -> false)
                    && laterFields
                       |> List.forall (fun field -> Set.contains field.Name earlierNames || not field.Required)
                | laterBody, earlierBody -> laterBody = earlierBody
            | _ -> false

File: Reference/src/Brontide.Reference.Experimental.Binding/Portable/PortableShapeCatalog.cs
using System.Collections.Immutable;

namespace Brontide.Reference.Experimental.Binding.Portable;

/// <summary>The Shape floor's built-in references, which no contract needs to redeclare.</summary>
public static class PortableBuiltInShapes
{
    public static PortableShapeReference Text { get; } = PortableShapeReference.Parse("Text", 1);
    public static PortableShapeReference Boolean { get; } = PortableShapeReference.Parse("Boolean", 1);
    public static PortableShapeReference Signed64 { get; } = PortableShapeReference.Parse("Integer.Signed64", 1);
    public static PortableShapeReference Decimal { get; } = PortableShapeReference.Parse("Decimal", 1);
    public static PortableShapeReference Bytes { get; } = PortableShapeReference.Parse("Bytes", 1);
    public static PortableShapeReference Unit { get; } = PortableShapeReference.Parse("Unit", 1);

    public static ImmutableArray<PortableShapeDeclaration> Declarations { get; } =
    [
        PortableShapeDeclaration.ScalarShape(Text, PortableScalarKind.Text),
        PortableShapeDeclaration.ScalarShape(Boolean, PortableScalarKind.Boolean),
        PortableShapeDeclaration.ScalarShape(Signed64, PortableScalarKind.Signed64),
        PortableShapeDeclaration.ScalarShape(Decimal, PortableScalarKind.Decimal),
        PortableShapeDeclaration.ScalarShape(Bytes, PortableScalarKind.Bytes),
        PortableShapeDeclaration.Unit(Unit)
    ];
}

/// <summary>
/// The Shape graph of one established contract, plus the projection rules that decide when a
/// version difference is additive.
/// </summary>
/// <remarks>
/// Unresolvable references fail as <c>unsupported-contract</c> rather than as a decode error,
/// because an unknown Shape, Fragment, or dependency is a contract fact, not a byte-level one.
/// </remarks>
public sealed class PortableShapeCatalog
{
    private readonly ImmutableDictionary<PortableShapeReference, PortableShapeDeclaration> _shapes;
    private readonly ImmutableDictionary<PortableFragmentReference, PortableFragmentDeclaration> _fragments;

    private PortableShapeCatalog(
        ImmutableDictionary<PortableShapeReference, PortableShapeDeclaration> shapes,
        ImmutableDictionary<PortableFragmentReference, PortableFragmentDeclaration> fragments,
        PortableLimits limits)
    {
        _shapes = shapes;
        _fragments = fragments;
        Limits = limits;
    }

    public PortableLimits Limits { get; }

    public static PortableShapeCatalog FromContract(PortableContractDocument document)
    {
        ArgumentNullException.ThrowIfNull(document);
        var shapes = ImmutableDictionary.CreateBuilder<PortableShapeReference, PortableShapeDeclaration>();
        foreach (var builtIn in PortableBuiltInShapes.Declarations)
        {
            shapes[builtIn.Reference] = builtIn;
        }

        foreach (var shape in document.Shapes)
        {
            if (shapes.ContainsKey(shape.Reference))
            {
                throw PortableFaultException.Malformed(
                    "duplicate-shape",
                    $"Shape {shape.Reference} is declared more than once.");
            }

            RequireUniqueFieldNames(shape.Fields, $"Shape {shape.Reference}");
            shapes[shape.Reference] = shape;
        }

        var fragments = ImmutableDictionary.CreateBuilder<PortableFragmentReference, PortableFragmentDeclaration>();
        foreach (var fragment in document.Fragments)
        {
            if (fragments.ContainsKey(fragment.Reference))
            {
                throw PortableFaultException.Malformed(
                    "duplicate-fragment",
                    $"Fragment {fragment.Reference} is declared more than once.");
            }

            RequireUniqueFieldNames(fragment.Fields, $"Fragment {fragment.Reference}");
            fragments[fragment.Reference] = fragment;
        }

        var catalog = new PortableShapeCatalog(shapes.ToImmutable(), fragments.ToImmutable(), document.Limits);
        catalog.RequireResolvable(document);
        return catalog;
    }

    public PortableShapeDeclaration Shape(PortableShapeReference reference) =>
        _shapes.TryGetValue(reference, out var declaration)
            ? declaration
            : throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "unknown-shape",
                $"Shape {reference} is not part of the established contract.");

    public bool TryShape(PortableShapeReference reference, out PortableShapeDeclaration declaration) =>
        _shapes.TryGetValue(reference, out declaration!);

    public PortableFragmentDeclaration Fragment(PortableFragmentReference reference) =>
        _fragments.TryGetValue(reference, out var declaration)
            ? declaration
            : throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "unknown-fragment",
                $"Fragment {reference} is not part of the established contract.");

    public bool TryFragment(PortableFragmentReference reference, out PortableFragmentDeclaration declaration) =>
        _fragments.TryGetValue(reference, out declaration!);

    /// <summary>
    /// Decides whether the later Shape version differs from the earlier one only additively: new
    /// optional fields. A new required field, a removed field, a changed field Shape, or a changed
    /// cardinality is not additive.
    /// </summary>
    public bool IsAdditiveOver(PortableShapeReference later, PortableShapeReference earlier)
    {
        if (later.Name != earlier.Name || later.Version < earlier.Version)
        {
            return false;
        }

        if (later == earlier)
        {
            return true;
        }

        if (!TryShape(later, out var laterShape) || !TryShape(earlier, out var earlierShape))
        {
            return false;
        }

        if (laterShape.Kind != earlierShape.Kind ||
            laterShape.FragmentPolicy != earlierShape.FragmentPolicy ||
            laterShape.ItemShape != earlierShape.ItemShape ||
            laterShape.Scalar != earlierShape.Scalar ||
            !laterShape.Alternatives.SequenceEqual(earlierShape.Alternatives))
        {
            return false;
        }

        var laterFields = laterShape.Fields.ToImmutableDictionary(field => field.Name, StringComparer.Ordinal);
        foreach (var field in earlierShape.Fields)
        {
            if (!laterFields.TryGetValue(field.Name, out var candidate) || candidate != field)
            {
                return false;
            }
        }

        var earlierNames = earlierShape.Fields.Select(field => field.Name).ToImmutableHashSet(StringComparer.Ordinal);
        return laterShape.Fields.All(field => earlierNames.Contains(field.Name) || !field.Required);
    }

    private void RequireResolvable(PortableContractDocument document)
    {
        foreach (var shape in document.Shapes)
        {
            foreach (var field in shape.Fields)
            {
                _ = Shape(field.Shape);
            }

            foreach (var alternative in shape.Alternatives)
            {
                _ = Shape(alternative.Shape);
            }

            if (shape.ItemShape is { } itemShape)
            {
                _ = Shape(itemShape);
            }
        }

        foreach (var fragment in document.Fragments)
        {
            _ = Shape(fragment.HostShape);
            foreach (var field in fragment.Fields)
            {
                _ = Shape(field.Shape);
            }
        }

        foreach (var operation in document.Operations)
        {
            _ = Shape(operation.InputShape);
            _ = Shape(operation.ResultShape);
            _ = Shape(operation.DetailShape);
            foreach (var fragment in operation.RequiredFragments)
            {
                _ = Fragment(fragment);
            }
        }
    }

    private static void RequireUniqueFieldNames(ImmutableArray<PortableFieldDeclaration> fields, string owner)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        foreach (var field in fields)
        {
            _ = PortableMemberToken.Parse(field.Name);
            if (!seen.Add(field.Name))
            {
                throw PortableFaultException.Malformed(
                    "duplicate-field",
                    $"{owner} declares field '{field.Name}' more than once.");
            }
        }
    }
}
